### ISSUE TITLE
production-tec AWS load balancer health check failure

### ISSUE THREAD

**Issue by cvallance** [coreos/tectonic-forum#74]
### Tectonic Version

1.5.2-tectonic.1

### Environment

AWS

### Expected Behavior

AWS load balancer for `production-tec` has healthy instances available... which in turn means I should be able to reach the cluster using kubectl and able to visit the tectonic console / dashboard.

### Actual Behavior

AWS load balancer for `production-tec` has no healthy instances available.

### More info

Cluster is still working fine, so the API for my app is still running (& backing database). Seems to be an issue with the worker but can't know if that's related or not.

It seems there was some erratic or spiking CPU on the machine just before it became unhealthy (via the `healthz` health check endpoint).

Anyway, I can still SSH into the machine, so here's a few things that might help diagnose the issue...

```
$ docker ps

```
The above `docker ps` simply hangs.

```
$ docker info
Containers: 64
 Running: 29
 Paused: 0
 Stopped: 35
Images: 26
Server Version: 1.12.6
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp selinux
Kernel Version: 4.9.9-coreos-r1
Operating System: Container Linux by CoreOS 1298.5.0 (Ladybug)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.862 GiB
Name: ip-10-0-0-50
ID: 7ZBV:CZKX:RWSQ:ZGTS:BTGA:OPZY:WAO3:Z5FS:FAUM:CKLV:OGNK:7VBA
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

```
$ systemctl status docker
● docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; enabled; vendor preset: enabled)
   Active: active (running) since Wed 2017-03-01 08:38:56 UTC; 1 weeks 1 days ago
     Docs: http://docs.docker.com
 Main PID: 1215 (dockerd)
    Tasks: 122
   Memory: 326.9M
      CPU: 1h 32min 653ms
   CGroup: /system.slice/docker.service
           └─1215 dockerd --host=fd:// --containerd=/var/run/docker/libcontainerd/docker-containerd.sock --selinux-enabled

Mar 08 00:30:16 ip-10-0-0-50 dockerd[1215]: time="2017-03-08T00:30:16.317434932Z" level=error msg="Handler for GET /v1.24/containers/json returned error: write unix /var/run/docker.sock->@: write: broken
Mar 08 00:30:16 ip-10-0-0-50 dockerd[1215]: http: multiple response.WriteHeader calls
Mar 08 00:30:18 ip-10-0-0-50 dockerd[1215]: time="2017-03-08T00:30:18.942782469Z" level=error msg="Handler for GET /v1.24/containers/b45bf774a9760e38623e557a995f5f1a3e46417ef32e6d3d2194c346e8622e37/json
Mar 08 00:30:42 ip-10-0-0-50 dockerd[1215]: time="2017-03-08T00:30:41.031449406Z" level=error msg="Handler for GET /containers/6db545fcace0502866fda3e01bddcf18e111d242020ae4c4fe7fb15127331fdd/json return
Mar 08 00:33:39 ip-10-0-0-50 dockerd[1215]: http: multiple response.WriteHeader calls
Mar 08 00:33:48 ip-10-0-0-50 dockerd[1215]: http: multiple response.WriteHeader calls
Mar 08 03:36:13 ip-10-0-0-50 dockerd[1215]: http: multiple response.WriteHeader calls
Mar 08 03:39:46 ip-10-0-0-50 dockerd[1215]: http: multiple response.WriteHeader calls
Mar 08 03:39:46 ip-10-0-0-50 dockerd[1215]: http: multiple response.WriteHeader calls
Mar 08 03:51:47 ip-10-0-0-50 dockerd[1215]: http: multiple response.WriteHeader calls
```

```
$ etcdctl cluster-health
member 8c76dbadd15001d1 is healthy: got healthy result from http://10.0.0.50:2379
cluster is healthy
```

I have read a little bit about docker becoming unresponsive when under load...? But what I did read says that it should have been fixed in the docker version Tectonic is running.

And really, it shouldn't be under that much load... aside from what comes bundled with Tectonic, I'm running 3 instances of a golang api, 1 instance of a golang worker and then 3 stateful set instances of mongo with a replica set sidecar.

Here's some images that may help pain the picture...

Load balancers showing the unhealthy instance:
<img width="848" alt="tecissue-loadbalancer" src="https://cloud.githubusercontent.com/assets/525090/23746131/b8471452-051f-11e7-87db-13f1ee0449b2.png">

Load balancer health over 3 days:
<img width="554" alt="screen shot 2017-03-09 at 11 24 33 pm" src="https://cloud.githubusercontent.com/assets/525090/23746153/cf9f1208-051f-11e7-9d5b-9e30133f027f.png">

EC2 Instances:
<img width="984" alt="tecissue-ec2instances" src="https://cloud.githubusercontent.com/assets/525090/23746350/7ba49f32-0520-11e7-9555-6d66ae6ab607.png">

EC2 Production Controller Instance CPU over 3 days:
<img width="275" alt="screen shot 2017-03-09 at 11 32 17 pm" src="https://cloud.githubusercontent.com/assets/525090/23746440/bcc510fa-0520-11e7-9f9f-209329ec912e.png">

**Comment by mfburnett** [comment]
Hi @cvallance, thanks for filing this issue. I've let @ivancherepov know. He'll get back to you in a few days and will pull in others from our team as necessary!

**Comment by IvanCherepov** [comment]
can you run `docker stats` and `sudo strace -f docker ps`(post as gist) on the instance?

**Comment by cvallance** [comment]
Thanks for looking into this @IvanCherepov... here's the info you requested.

`docker stats` hangs like `docker ps` does... 😞 

`sudo strace -f docker ps`
https://gist.github.com/cvallance/e6a283c477e5a990eafd8c6757db7bc1

`sudo strace -f docker stats`
https://gist.github.com/cvallance/f586f5c8afeb8f3a7367ce17b06c387a

**Comment by IvanCherepov** [comment]
could you also run `sudo kill -SIGUSR1 $(cat /run/docker.pid)`(it will send SIGUSR1 signal to the docker daemon process) and put the result of `sudo journalctl -u docker -l` into a gist?

**Comment by cvallance** [comment]
Done. Gist coming soon... it's 14mb.

**Comment by cvallance** [comment]
@IvanCherepov - gist for `sudo journalctl -u docker -l`
https://gist.github.com/cvallance/d6c63450c474f23e247b4086b8acd6ab

What does the `SIGUSR1` do for docker?

**Comment by IvanCherepov** [comment]
Sending SIGUSR1 signal dumps docker daemon stack trace into logs. There are a lot of "No such container", "rpc error: code = 2" and "rpc error: code = 4"  errors.

**Comment by cvallance** [comment]
Yeah, it's basically all `no such container` errors. If I reboot the docker daemon, would that fix things? Would I lose anything?

Could I put this into `/etc/docker/daemon.json` and then do a `kill -HUP $(cat /var/run/docker.pid)`
```
{
    "debug": true,
    "live-restore": true
}
```
ala https://github.com/docker/docker/issues/25381#issuecomment-238602268

As I say, the containers for my application are still running and are currently working (but I can't update the deployments via kubectl)... so I'm happy if you want me to keep it in this state for a little bit so you can diagnose the issue to fix a bug or maybe help others down the road.

**Comment by dghubble** [comment]
How many controllers are running (they back `production-k8s` and `production-tec` ELBs)?

In Tectonic 1.5.2-tectonic.2 and higher, we started deploying etcd separately from the controller so an easy solution would be to scale the controller auto-scaling group and repair/delete the unhealthy one (depending on how much you care).

I've had luck restarting the docker service to resolve `docker ps` hanging on a few occasions.

**Comment by cvallance** [comment]
@dghubble There's only 1 controller (+ 2 workers) ... the only autoscaling group I have is for the workers, there wasn't one setup for the controller(s).

I'll restart the docker service in the next hour or so unless anyone suggests otherwise. Cheers.

**Comment by cvallance** [comment]
Restarting the docker service resolved the issue and there didn't seem to be any issues or loss of data. So I'll just do that if it happens again.

Not sure if there is some underlying issue that needs more attention, but I'll try and update this issue with info if it happens again.

Feel free to close this issue if you wish. Cheers.

**Comment by IvanCherepov** [comment]
Thank you for reporting back! I'll ask another engineer to take  a look at logs and will post an update in case he'll find something.

**Comment by mfburnett** [comment]
@cvallance going to close this issue - feel free to reopen with further comments. We recently released Tectonic 1.6.2-tectonic.1 that you should check out.
